aarch64/src/interrupts.S

```diff
@@ -0,0 +1,170 @@
+.section .text
+
+.equ	SYNC_INVALID_EL1t,	0
+.equ	IRQ_INVALID_EL1t,	1
+.equ	FIQ_INVALID_EL1t,	2
+.equ	ERROR_INVALID_EL1t,	3
+
+.equ	SYNC_INVALID_EL1h,	4
+.equ	IRQ_INVALID_EL1h,	5
+.equ	FIQ_INVALID_EL1h,	6
+.equ	ERROR_INVALID_EL1h,	7
+
+.equ	SYNC_INVALID_EL0_64,	8
+.equ	IRQ_INVALID_EL0_64,	9
+.equ	FIQ_INVALID_EL0_64,	10
+.equ	ERROR_INVALID_EL0_64,	11
+
+.equ	SYNC_INVALID_EL0_32,	12
+.equ	IRQ_INVALID_EL0_32,	13
+.equ	FIQ_INVALID_EL0_32,	14
+.equ	ERROR_INVALID_EL0_32,	15
+
+.macro	ventry	label
+.align	7
+	b	\label
+.endm
+
+// Aarch64 calling convention dictates that we save:
+// - x0-18
+// - x29 (Frame pointer)
+// - x30 (Link register)
+// We additionally save the registers below, so that they're avaialble via the InterruptFrame
+// - ESR_EL1 (Exception syndrome register EL1)
+// - ELR_EL1 (Exception link register EL1)
+// - FAR_EL1 (Fault address register EL1)
+.macro handle_interrupt type
+	sub 	sp, sp, #208
+
+	// Caller-saved registers, FP
+	stp 	x0, x1, [sp, #16 * 0]
+	stp 	x2, x3, [sp, #16 * 1]
+	stp	x4, x5, [sp, #16 * 2]
+	stp	x6, x7, [sp, #16 * 3]
+	stp	x8, x9, [sp, #16 * 4]
+	stp	x10, x11, [sp, #16 * 5]
+	stp	x12, x13, [sp, #16 * 6]
+	stp	x14, x15, [sp, #16 * 7]
+	stp	x16, x17, [sp, #16 * 8]
+	stp	x18, x29, [sp, #16 * 9]
+
+	// LR, ESR_EL1
+	mrs	x0, esr_el1
+	stp	x30, x0, [sp, #16 * 10]
+
+	// ELR_EL1, FAR_EL1
+	mrs	x1, elr_el1
+	mrs	x2, far_el1
+	stp	x1, x2, [sp, #16 * 11]
+
+	// Interrupt type, padding
+	ldr	x3, =\type
+	stp	x3, xzr, [sp, #16 * 12]
+
+	// Pass pointer to InterruptFrame (on stack) as the first arg
+	mov	x0, sp
+	bl	handle_exception_with_ptr
+
+	// Restore caller-saved registers
```

Comment:
Just a point to note that, on interrupt, unlike on (say) a context switch, you've got to save all registers. This looks like what's going on, though?

Reply:
I was assuming the calling of the rust function handle_exception_with_ptr would cover the callee-saved registers.

Reply:
One must be very careful here, as it could be a potential way to leak things across processes by accident. Moreover, often times one wants to modify register contents on return; if we don't have them in the context structure, we can't do that. :-/

I'd also call the Rust function `trap`, as that's the plan9 name.

Reply:
ok, that makes sense - I'll make the change.